polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -3314,10 +3316,19 @@ void MLIRASTConsumer::createMLIRParametersTypes(
     parmTypes.push_back(mlirThisType);
   }
 
+  const bool isKernel = FD.hasAttr<SYCLKernelAttr>();
+
   // Handle the remaining parameters.
   for (ParmVarDecl *parm : FD.parameters()) {
     QualType parmType = parm->getType();
 
+    // SPIR ABI compliance: aggregates need to be passed as pointers.
+    if (isKernel && CodeGenUtils::isAggregateTypeForABI(parmType)) {
+      auto mlirType = mlir::MemRefType::get(-1, getMLIRType(parmType));
```

Comment:
Should the address space for this be `4`?

Reply:
We should use the same addr. space clang uses for "indirect" arguments (clang classifies kernel args as indirect). That is, we should use alloca addr space (see CGCall.cpp line 1702).

Reply:
Sorry to come back to this topic. But, if we use SPIRV's private address space here, we'll run into issues when kernels are to be run in a device. I'd say we should use SPIRV's generic address space (4) instead.

Reply:
We are using the same address space that clang uses for SPIR kernels. Given it works for SYCL device code compiled using clang (when clang generates LLVM IR) I would like to do the same when `cgeist` is called.

Reply:
I misread the condition. LGTM